File: C++/ScientificCalculator/src/Engine.cpp
#include "Engine.h"

#include "FunctionLibrary.h"
#include "Tokenizer.h"
#include "Parser.h"

#include <vector>
#include <string>


using namespace std;



// =======================
// Initialize
// =======================

void Engine::initialize()
{

    static bool initialized =
        false;


    if(initialized)
        return;


    FunctionLibrary::initialize();


    initialized = true;

}



// =======================
// Evaluate
// =======================

double Engine::evaluate(
    const string& expression
)
{

    initialize();


    vector<string> tokens =
        tokenize(
            expression
        );


    Parser parser(
        tokens
    );


    return parser.evaluate();

}

File: Java/ScientificCalculator/src/Engine.java
import java.util.List;

import parser.Tokenizer;
import parser.Parser;


public class Engine {


    // ==========================
    // Evaluate Expression
    // ==========================

    public static double evaluate(
            String expression
    ) {


        if (
            expression == null
            ||
            expression.trim().isEmpty()
        ) {

            throw new IllegalArgumentException(
                "Expression is empty"
            );

        }



        try {


            List<String> tokens =
                    Tokenizer.tokenize(
                        expression
                    );



            Parser parser =
                    new Parser(
                        tokens
                    );



            return parser.parse();


        }


        catch (NumberFormatException e) {


            throw new IllegalArgumentException(
                "Invalid number format"
            );


        }


        catch (IndexOutOfBoundsException e) {


            throw new IllegalArgumentException(
                "Incomplete expression"
            );


        }

    }

}
